Remove the second copy on calling dist_autograd_context._known_worker_ids()

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31206 Remove the second copy on calling dist_autograd_context._known_worker_ids()**

Improvement on #25525.

- DistAutogradContext::getKnownWorkerIds() returns a unordered_map as temp value. No need to copy this temp value A into another temp value B.

Differential Revision: [D5707771](https://our.internmc.facebook.com/intern/diff/D5707771/)

xush6528 committed:
Remove the second copy on calling dist_autograd_context._known_worker_ids()

Improvement on #25525.

- DistAutogradContext::getKnownWorkerIds() returns a unordered_map as temp value. No need to copy this temp value A into another temp value B.

Differential Revision: [D5707771](https://our.internmc.facebook.com/intern/diff/D5707771/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Thanks! One thing is that the return object will now be an `unordered_map` instead of a `vector`, so I think a Python `set()` would be returned when using this method. I think the only callsites are in tests so this should be fine as long as those tests pass

kostmo commented:
## CircleCI build failures summary
As of commit 6c20039:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6c2003960973f0a002ced7cf8d8ec5c0af6de1c8).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3930473) pytorch-linux-xenial-py3-clang5-android-ndk-r19c-gradle-build (1/2)
**Step:** "pytorch android gradle build" ([details](https://dr.pytorch.org/build-details.html?build_id=73554142))
```
sys	0m0.091s 
++ docker run --cap-add=SYS_PTRACE --security-opt seccomp=unconfined -t -d -w /var/lib/jenkins 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-6c2003960973f0a002ced7cf8d8ec5c0af6de1c8-android-x86_32 
+ export id_x86_32=e5a06a4adf9f26700f782d6ea841efce625193dc45b66209b638423a36dc5da5 
+ id_x86_32=e5a06a4adf9f26700f782d6ea841efce625193dc45b66209b638423a36dc5da5 
+ export 'COMMAND=((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ COMMAND='((echo "source ./workspace/env" && echo "sudo chown -R jenkins workspace") | docker exec -u jenkins -i "$id_x86_32" bash) 2>&1' 
+ echo '((echo' '"source' './workspace/env"' '&&' echo '"sudo' chown -R jenkins 'workspace")' '|' docker exec -u jenkins -i '"$id_x86_32"' 'bash)' '2>&1' 
+ unbuffer bash ./command.sh 
+ ts 
+ docker pull 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-6c2003960973f0a002ced7cf8d8ec5c0af6de1c8-android-arm-v7a 
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-6c2003960973f0a002ced7cf8d8ec5c0af6de1c8-android-arm-v7a not found 
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3929704) pytorch_linux_xenial_py3_6_gcc5_4_test (2/2)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73554206))
```
Dec 12 21:29:15 Ran 48 tests in 67.477s 
Dec 12 21:29:15  
Dec 12 21:29:15 FAILED (failures=2, skipped=2) 
Dec 12 21:29:15  
Dec 12 21:29:15 Generating XML reports... 
Dec 12 21:29:15 Traceback (most recent call last): 
Dec 12 21:29:15   File "test/run_test.py", line 456, in <module> 
Dec 12 21:29:15     main() 
Dec 12 21:29:15   File "test/run_test.py", line 449, in main 
Dec 12 21:29:15     raise RuntimeError(message) 
Dec 12 21:29:15 RuntimeError: test_dist_autograd_spawn failed! 
Dec 12 21:29:15 + cleanup 
Dec 12 21:29:15 + retcode=1 
Dec 12 21:29:15 + set +x 
Dec 12 21:29:15 =================== sccache compilation log =================== 
Dec 12 21:29:15 ERROR:sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "/tmp/torch_extensions/test_compilation_error_formatting/main.cpp: In function \'int main()\':\n/tmp/torch_extensions/test_compilation_error_formatting/main.cpp:2:23: error: expected \';\' before \'}\' token\n int main() { return 0 }\n                       ^\n" } 
Dec 12 21:29:15  
Dec 12 21:29:15 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Dec 12 21:29:15 Compile requests                115 
Dec 12 21:29:15 Compile requests executed        25 
Dec 12 21:29:15 Cache hits                       18 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

xush6528 committed:
Update on "Remove the second copy on calling dist_autograd_context._known_worker_ids()"


Improvement on #25525.

- DistAutogradContext::getKnownWorkerIds() returns a unordered_map as temp value. No need to copy this temp value A into another temp value B.

Differential Revision: [D5707771](https://our.internmc.facebook.com/intern/diff/D5707771/)

[ghstack-poisoned]

xush6528 commented:
@rohan-varma 

Thanks for the reminder. I updated the tests.

xush6528 committed:
Update on "Remove the second copy on calling dist_autograd_context._known_worker_ids()"


Improvement on #25525.

- DistAutogradContext::getKnownWorkerIds() returns a unordered_map as temp value. No need to copy this temp value A into another temp value B.

Differential Revision: [D5707771](https://our.internmc.facebook.com/intern/diff/D5707771/)

[ghstack-poisoned]

rohan-varma commented on test/dist_autograd_test.py:
```diff
@@ -99,19 +99,25 @@ def my_py_nested_call(t1, t2, dst, world_size, hops):
 # nodes. This helper allows timeout_seconds for those RPCs to be completed, and
 # ensures that all the contexts have been cleaned up in that timeframe.any
 def _all_contexts_cleaned_up(timeout_seconds=10):
+    assert dist.is_initialized(), "ProcessGroup must be initialized."
```
Hmm, I think its better for the caller test methods to call `barrier()` on their own, as in some cases, the barrier is not necessary (see `test_context_cleanup_many_workers` and `test_context_cleanup_tensor_no_grad`). I'm not sure if we'd want to add `dist.barrier()` calls unless it is strictly necessary for the test

xush6528 replied:
@rohan-varma Moved `barrier()` back to the 3 tests where I noticed flakiness.

xush6528 committed:
Update on "Remove the second copy on calling dist_autograd_context._known_worker_ids()"


Improvement on #25525.

- DistAutogradContext::getKnownWorkerIds() returns a unordered_map as temp value. No need to copy this temp value A into another temp value B.

Differential Revision: [D5707771](https://our.internmc.facebook.com/intern/diff/D5707771/)

[ghstack-poisoned]
